cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/JmsSource.java

```diff
@@ -0,0 +1,219 @@
+ * </p>
+ */
+public class JmsSource extends RealtimeSource<String> implements MessageListener {
+  private static final Logger LOG = LoggerFactory.getLogger(JmsSource.class);
+
+  private int jmsAcknowledgeMode = Session.AUTO_ACKNOWLEDGE;
+  private JmsProvider jmsProvider;
+  private final LinkedBlockingQueue<Message> messageQueue = new LinkedBlockingQueue<Message>();
+
+  private transient Connection connection;
+  private transient Session session;
+
+  /**
+   * Configure the JMS Source.
+   *
+   * @param configurer {@link RealtimeConfigurer}
+   */
+  @Override
+  public void configure(RealtimeConfigurer configurer) {
+    configurer.setName("JMS Realtime Source");
+    configurer.setDescription("CDAP JMS Realtime Source");
+  }
+
+  /**
+   * Initialize the Source.
+   *
+   * @param context {@link SourceContext}
+   */
+  public void initialize(SourceContext context) {
+    super.initialize(context);
+
+    // Bootstrap the JMS consumer
+    initializeJMSConnection();
+  }
+
+  /**
+   * Helper method to initialize the JMS Connection to start listening messages.
+   */
+  private void initializeJMSConnection() {
+    if(jmsProvider == null) {
+      throw new IllegalStateException("Could not have null JMSProvider for JMS Source. " +
+                                         "Please set the right JMSProvider");
+    }
+    ConnectionFactory connectionFactory = jmsProvider.getConnectionFactory();
+
+    try {
+      connection = connectionFactory.createConnection();
+      session = connection.createSession(false, jmsAcknowledgeMode);
+      Destination destination = jmsProvider.getDestination();
+      MessageConsumer consumer = session.createConsumer(destination);
+      consumer.setMessageListener(this);
+      connection.start();
+    } catch (JMSException ex) {
+      if(session != null) {
+        try {
+          session.close();
+        } catch (JMSException ex1) {
+          LOG.debug("Exception when closing session", ex1);
+        }
+      }
+      if (connection != null) {
+        try {
+          connection.close();
+        } catch (JMSException ex2) {
+          LOG.debug("Exception when closing connection", ex2);
+        }
+      }
+      throw new RuntimeException("JMSException thrown when trying to initialize connection: " + ex.getMessage(),
+                                 ex);
+    }
+  }
+
+  @Nullable
+  @Override
+  public SourceState poll(ValueEmitter<String> writer, SourceState currentState) {
+    // Try to get message from Queue
+    Message message = messageQueue.poll();
+    if (message == null) {
+      return currentState;
+    }
+
+    if (message instanceof TextMessage) {
+      TextMessage textMessage = (TextMessage) message;
+      try {
+        String text = textMessage.getText();
+        LOG.debug("Process JMS TextMessage : " + text);
+        writer.emit(text);
+      } catch (JMSException e) {
+        LOG.error("Unable to read text from a JMS TextMessage.");
```

Comment:
Ah, I am assuming error level is used for non recoverable problems which I assume include unable to read message from the valid JMS TextMessage.